### PR TITLE
ThunderAgent integration and R2Egym code agent rl training receipe.

### PR DESCRIPTION
**Pull request by HaoKang-Timmy**
## Summary

  This PR adds a ThunderAgent + SkyRL training recipe for fully async Harbor/R2EGym training with program-aware inference scheduling.

  The example lives under:

  - `examples/train/thunder_agent/`

  It includes:
  - a Harbor/R2EGym Qwen3-32B training entrypoint;
  - ThunderAgent-aware SkyRL inference routing;
  - rollout program release hooks;
  - weight-sync coordination around inference pause/resume;
  - Slurm scripts for the 5-node R2EGym 32B recipe;
  - dataset preparation and operational helpers;
  - router and scheduler regression tests;
  - documentation for the benchmark recipe and reproducible data split.

  ## Motivation

  ThunderAgent improves agent RL rollout throughput by scheduling inference at the program level rather than treating every request
  independently. This is especially useful for multi-turn agent workloads where trajectories have heterogeneous context lengths and
  execution times.

  This example demonstrates the integration on Harbor/R2EGym and documents the benchmark configuration used for the reported 40-step
  run.

  ## Key Files

  - `examples/train/thunder_agent/README.md`
    - End-to-end recipe documentation, hardware layout, data prep, launch flow, validation, and tunable parameters.

  - `examples/train/thunder_agent/main_harbor_thunder_agent.py`
    - Main fully async Harbor training entrypoint.

  - `examples/train/thunder_agent/main_thunder_agent.py`
    - Shared ThunderAgent experiment base that swaps SkyRL’s standard HTTP inference router for `ThunderAgentRouter`.

  - `examples/train/thunder_agent/skyrl_integration/router.py`
    - SkyRL integration router that composes ThunderAgent routes with SkyRL’s `/inference/v1/generate` and `/servers` endpoints.

  - `examples/train/thunder_agent/skyrl_integration/remote_inference_client.py`
    - Remote inference client wrapper that brackets pause/resume with ThunderAgent weight-sync begin/end calls.

  - `examples/train/thunder_agent/skyrl_integration/generator.py`
    - Harbor generator adapter that attaches routing IDs, releases ThunderAgent programs after trajectories, and preserves rollout
  logprobs when requested.

  - `examples/train/thunder_agent/scripts/r2egym_32b/`
    - Slurm and operational scripts for the R2EGym 32B recipe.

  - `tests/backends/skyrl_train/inference_servers/test_thunder_agent_router.py`
    - Router, program tracking, weight-sync, metadata stripping, and scheduler regression tests.

  ## Notes For Reviewers

  The recipe uses the new SkyRL HTTP inference path and requires:

  ```bash
  _SKYRL_USE_NEW_INFERENCE=1

  The default documented benchmark layout is:

  - 1 merged head/rollout node with 8 H100s
      - Ray head
      - training driver
      - ThunderAgent proxy
      - 4 vLLM rollout servers at TP=2
  - 4 trainer nodes with 8 H100s each
      - FSDP2 policy/ref training

  Total: 5 nodes, 40 H100 GPUs.

  The integration supports both:

  - embedded ThunderAgent router over external rollout servers;
  - externally provided ThunderAgent proxy via EXTERNAL_PROXY_URL / THUNDERAGENT_URL.

### CODE REVIEW

File: examples/train/thunder_agent/README.md [hunk 1]
```diff
@@ -0,0 +1,247 @@
+# ThunderAgent + SkyRL: R2EGym 32B Training Recipe
+
+Train Qwen3-32B on R2EGym with ThunderAgent-accelerated rollout scheduling.
+
+![TA vs no-TA stitched 40-step Harbor timeline](./docs/speedup.png)
+
+**3.01× wall-clock speedup** over the no-TA baseline (8.84 h vs 26.58 h for 40 training steps / 10 epochs).
```
Comment (CharlieFRuan) [inline examples/train/thunder_agent/README.md:7]: Can we add something like

```
This example uses SkyRL + Harbor integration but with MiniSWEAgent rather than Terminus 2. Therefore it does not obey token-in-token-out (`examples/train_integrations/harbor` does). This ThunderAgent example is mainly for demonstrating speedup rather than an actual RL recipe. The same ThunderAgent setup should be able to extend to Terminus 2 in Harbor, or other agent harness.
```

File: examples/train/thunder_agent/tests/test_thunder_agent_router.py
Comment (CharlieFRuan) [inline examples/train/thunder_agent/tests/test_thunder_agent_router.py:1]: Could we move this `test_thunder_agent_router.py` to `examples/train/thunder_agent/tests` please?

File: .codex
Comment (CharlieFRuan) [inline .codex:1]: let's remove this file

File: examples/train/thunder_agent/README.md [hunk 2]
```diff
@@ -0,0 +1,247 @@
...
+
+---
+
+## Hardware
+
+| Role | Nodes | GPUs | Notes |
+|---|---|---|---|
+| Head + rollout | 1 | 8 × H100 | Ray head, ThunderAgent proxy, training driver, 4 vLLM servers at TP=2 |
+| Trainer | 4 | 8 × H100 each | FSDP2 policy + ref model |
```
Comment (CharlieFRuan) [inline examples/train/thunder_agent/README.md:21]: Hmm is there a reason why we would want 4 training nodes and 1 rollout node? Usually it might be the other way around (e.g. 3 rollout nodes and 1 training node)

**Review by CharlieFRuan — COMMENTED** [review]
Thank you! Left various comments!

**Review by CharlieFRuan — APPROVED** [review]
LGTM, thanks!
